src/metabase/models/session.clj

```diff
@@ -29,3 +39,68 @@
+  (when (and (login-history/send-email-on-first-login-from-new-device)
+             (login-history/first-login-on-this-device? history-record)
+             (not (login-history/first-login-ever? history-record)))
+    ;; if there's an existing open connection (and there seems to be one, but I'm not 100% sure why) we can't try to use
+    ;; it across threads since it can close at any moment! So unbind it so the future can get its own thread.
+    (binding [t2.conn/*current-connectable* nil]
+      (future
+        ;; off thread for both IP lookup and email sending. Either one could block and slow down user login (#16169)
+        (try
+          (let [[info] (login-history/human-friendly-infos [history-record])]
+            (messages/send-login-from-new-device-email! info))
+          (catch Throwable e
+            (log/error e "Error sending 'login from new device' notification email")))))))
+
+(def ^:private CreateSessionUserInfo
+  [:map
+   [:id ms/PositiveInt]
+   [:last_login :any]])
+
+(def SessionSchema
+  "Schema for a Session."
+  [:and
+   [:map-of :keyword :any]
+   [:map
+    [:id uuid?]
+    [:type [:enum :normal :full-app-embed]]]])
+
+(defmulti create-session!
+  "Generate a new Session for a User. `session-type` is currently either `:password` (for email + password login) or
+  `:sso` (for other login types). Returns the newly generated Session."
+  {:arglists '(^java.util.UUID [session-type user device-info])}
+  (fn [session-type & _]
+    session-type))
+
+(mu/defmethod create-session! :sso :- SessionSchema
+  [_ user :- CreateSessionUserInfo device-info :- request/DeviceInfo]
+  (let [session-id (random-uuid)
+        session (first (t2/insert-returning-instances! :model/Session
+                                                       :id (str session-id)
+                                                       :user_id (u/the-id user)))]
+    (assert (map? session))
+    (let [event {:user-id (u/the-id user)}]
+      (events/publish-event! :event/user-login event)
+      (when (nil? (:last_login user))
+        (events/publish-event! :event/user-joined event)))
+    (let [history-entry (login-history/record-login-history! session-id (u/the-id user) device-info)]
+      (when-not (:embedded device-info)
+        (maybe-send-login-from-new-device-email history-entry))
+      (when-not (:last_login user)
+        (snowplow/track-event! ::snowplow/account {:event :new-user-created} (u/the-id user)))
+      (assoc session :id session-id))))
+
+(mu/defmethod create-session! :password :- SessionSchema
+  [session-type
+   user :- CreateSessionUserInfo
+   device-info :- request/DeviceInfo]
+  ;; this is actually the same as `create-session!` for `:sso` but we check whether password login is enabled.
+  (when-not (public-settings/enable-password-login)
+    (throw (ex-info (str (tru "Password login is disabled for this instance.")) {:status-code 400})))
+  ((get-method create-session! :sso) session-type user device-info))
```

Comment:
Right now, the `:sso` implementation is kind of the base implementation, but I wonder if it makes sense to make this more explicit? E.g. someone could decide "hey, I'll save the session type as part of the login history" and then we'd accidentally record `:password` logins as `:sso` logins. On the other hand this is small enough right now to see at a glance what's happening, so don't let this block you.

Reply:
I think I'll leave it as-is for now, and as the code gets more complex we can think about it